File: src/stores/history.ts
/** Download history commands; SQLite and transactions belong to Rust. */
import { defineStore } from 'pinia'
import { invoke } from '@tauri-apps/api/core'
import { useDatabaseStore } from '@/stores/database'
import type { HistoryRecord } from '@shared/types'

export type HistoryRecordSortField = 'name' | 'status' | 'total_length' | 'task_type' | 'completed_at'
export type HistoryRecordSortOrder = 'ascend' | 'descend' | false

export interface HistoryRecordsPageInput {
  status?: string
  page: number
  pageSize: number
  sortField?: string
  sortOrder?: HistoryRecordSortOrder
}

export interface HistoryRecordsPage {
  records: HistoryRecord[]
  total: number
}

export const useHistoryStore = defineStore('history', () => {
  const database = useDatabaseStore()
  async function call<T>(command: string, args?: Record<string, unknown>): Promise<T> {
    await database.init()
    return invoke<T>(command, args)
  }
  return {
    init: database.init,
    addRecord: (record: HistoryRecord) => call<void>('history_add_record', { record }),
    getRecords: (status?: string, limit?: number) => call<HistoryRecord[]>('history_get_records', { status, limit }),
    getRecordsPage: (input: HistoryRecordsPageInput) =>
      call<HistoryRecordsPage>('history_get_page', {
        input: { ...input, sortOrder: input.sortOrder || null },
      }),
    getRecordByGid: (gid: string) => call<HistoryRecord | null>('history_get_record', { gid }),
    removeRecord: (gid: string) => call<void>('history_remove_record', { gid }),
    removeBirthRecords: (gids: string[]) => call<void>('history_remove_births', { gids }),
    clearRecords: (status?: string) => call<void>('history_clear_records', { status }),
    removeStaleRecords: (gids: string[]) => call<void>('history_remove_stale', { gids }),
    removeByInfoHash: (infoHash: string, excludeGid?: string) =>
      call<void>('history_remove_by_info_hash', { infoHash, excludeGid }),
    checkIntegrity: () => call<string>('history_check_integrity'),
    recordTaskBirth: (gid: string, addedAt = new Date().toISOString()) =>
      call<void>('history_record_birth', { gid, addedAt }),
    async loadBirthRecords(): Promise<Array<{ gid: string; added_at: string }>> {
      const records = await call<Array<[string, string]>>('history_load_births')
      return records.map(([gid, added_at]) => ({ gid, added_at }))
    },
    getSchemaVersion: () => call<number>('database_schema_version'),
  }
})

File: src-tauri/src/commands/history.rs
//! History commands backed by process-owned SQLite storage.
use crate::database::{DatabaseState, HistoryPage, HistoryPageInput, HistoryRecord};
use crate::error::AppError;
use tauri::State;

/// Add or update a history record (upsert by GID).
#[tauri::command]
pub async fn history_add_record(
    state: State<'_, DatabaseState>,
    record: HistoryRecord,
) -> Result<(), AppError> {
    state.0.add_record(&record).await
}

/// Query history records, optionally filtered by status and limited.
#[tauri::command]
pub async fn history_get_records(
    state: State<'_, DatabaseState>,
    status: Option<String>,
    limit: Option<u32>,
) -> Result<Vec<HistoryRecord>, AppError> {
    state.0.get_records(status.as_deref(), limit).await
}

/// Remove a single record by GID.
#[tauri::command]
pub async fn history_remove_record(
    state: State<'_, DatabaseState>,
    gid: String,
) -> Result<(), AppError> {
    state.0.remove_record(&gid).await
}

/// Clear records, optionally filtered by status.
#[tauri::command]
pub async fn history_clear_records(
    state: State<'_, DatabaseState>,
    status: Option<String>,
) -> Result<(), AppError> {
    state.0.clear_records(status.as_deref()).await
}

/// Remove records whose GIDs are in the provided list.
#[tauri::command]
pub async fn history_remove_stale(
    state: State<'_, DatabaseState>,
    gids: Vec<String>,
) -> Result<(), AppError> {
    state.0.remove_stale_records(&gids).await
}

/// Remove records matching a BT infoHash in the meta JSON column.
#[tauri::command]
pub async fn history_remove_by_info_hash(
    state: State<'_, DatabaseState>,
    info_hash: String,
    exclude_gid: Option<String>,
) -> Result<(), AppError> {
    state
        .0
        .remove_by_info_hash(&info_hash, exclude_gid.as_deref())
        .await
}

/// Record a task birth timestamp.
#[tauri::command]
pub async fn history_record_birth(
    state: State<'_, DatabaseState>,
    gid: String,
    added_at: String,
) -> Result<(), AppError> {
    state.0.record_task_birth(&gid, &added_at).await
}

/// Load all birth records.
#[tauri::command]
pub async fn history_load_births(
    state: State<'_, DatabaseState>,
) -> Result<Vec<(String, String)>, AppError> {
    state.0.load_birth_records().await
}

/// Run PRAGMA integrity_check.
#[tauri::command]
pub async fn history_check_integrity(state: State<'_, DatabaseState>) -> Result<String, AppError> {
    state.0.check_integrity().await
}

#[tauri::command]
pub async fn history_get_record(
    state: State<'_, DatabaseState>,
    gid: String,
) -> Result<Option<HistoryRecord>, AppError> {
    state.0.get_record(&gid).await
}
#[tauri::command]
pub async fn history_get_page(
    state: State<'_, DatabaseState>,
    input: HistoryPageInput,
) -> Result<HistoryPage, AppError> {
    state.0.get_records_page(input).await
}
#[tauri::command]
pub async fn history_remove_births(
    state: State<'_, DatabaseState>,
    gids: Vec<String>,
) -> Result<(), AppError> {
    state.0.remove_birth_records(&gids).await
}
